ibis/expr/operations/udf.py

```diff
@@ -51,21 +57,130 @@ def _wrap(
     return functools.update_wrapper(wrapper(input_type, fn, *args, **kwargs), fn)
 
 
-S = TypeVar("S", bound=ScalarUDF)
+S = TypeVar("S", bound=ops.Value)
+B = TypeVar("B", bound=ops.Value)
+
+
+class _UDF(abc.ABC):
+    __slots__ = ()
+
+    @property
+    @abc.abstractmethod
+    def base(self) -> type[B]:
```

Comment:
Perhaps make `.base` a private property (so it's not available as `ibis.udf.scalar.base`).

Reply:
Done, it's now `_base`